### PR TITLE
Building a TensorProperty AbstractBaseClass

### PR DESCRIPTION
**Pull request by Gamrix**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69802
* #70209
* #69051
* #69050
* __->__ #71184
* #69052
* #69049

Differential Revision: [D33555890](https://our.internmc.facebook.com/intern/diff/D33555890)

### CODE REVIEW

File: torch/csrc/jit/passes/shape_analysis.h [hunk 1]
```diff
@@ -9,6 +9,28 @@ namespace jit {
...
+  virtual ~PropertyPropBase() = default;
+
+  void propagateBlock(Block* block, bool insert_expands = true);
+  // insert_expands is used for shape inference
+
+  void processIf(Node* node);
+
+  protected:
+  virtual void propagateNode(Node* node, bool insert_expands = true) = 0;
+  void setUnshapedType(Value* o);
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/shape_analysis.h:29]: I guess we could make these `virtual` if we need to later

File: torch/csrc/jit/passes/shape_analysis.h [hunk 2]
```diff
@@ -9,6 +9,28 @@ namespace jit {
...
+
+class PropertyPropBase{
+  // Used for both Shape Propagation and Dtype/Device Propagation
+  public:
+  explicit PropertyPropBase(std::shared_ptr<Graph> graph)
+      : graph_(std::move(graph)) {
+  }
+  virtual ~PropertyPropBase() = default;
+
+  void propagateBlock(Block* block, bool insert_expands = true);
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/shape_analysis.h:22]: this isn't great, but I think I like the alternatives even less like 
a) moving `insert_expands` to be the part of analysis' state
b) duplicating/specializing `processLoop` for Shape Analysis vs Dtype and Device prop
c) splitting `propagateBlock(..., insert_expands)` into `propagateBlock` and `postPropagateBlock` 
or `propagateBlockLoop` and `propagateBlock`

@eellison wdyt ?

**Review by Krovatkin — APPROVED** [review]
:shipit:

**Comment by Gamrix** [comment]
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33555890).

**Pull request closed without merging**